src/client/texture_atlas.cpp

```diff
@@ -0,0 +1,236 @@
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "client/texture_atlas.h"
+#include "settings.h"
+#include "log.h"
+
+
+TextureAtlas::TextureAtlas(video::IVideoDriver *vdrv, ITextureSource *src, std::vector<TileLayer*> &layers)
+{
+	m_driver = vdrv;
+
+	m_tsrc = src;
+
+	u32 atlas_size = 0;
+
+	// Tile layers with unique ITextures
+	std::vector<TileLayer *> unique_layers;
+
+	for (auto layer : layers) {
+		// If for the layer the TileInfo was already created,
+		// just save the index at this TileInfo and omit the iteration
+		bool is_layer_collected = false;
+		for (u32 i = 0; i < unique_layers.size(); i++)
+			if (unique_layers[i]->texture == layer->texture) {
+				layer->atlas_tile_info_index = i;
+				is_layer_collected = true;
+				break;
+			}
+
+		if (is_layer_collected)
+			continue;
+
+		TileInfo tile_info;
+
+		if (layer->animation_frame_count > 1)
+			tile_info.tex = (*layer->frames)[0].texture;
+		else
+			tile_info.tex = layer->texture;
+
+		if (tile_info.tex) {
+			core::dimension2du size = tile_info.tex->getSize();
+			tile_info.width = size.Width;
+			tile_info.height = size.Height;
+		}
+
+		if (layer->animation_frame_count > 1) {
+			tile_info.anim.frame_length_ms = layer->animation_frame_length_ms;
+			tile_info.anim.frame_count = layer->animation_frame_count;
+
+			//if (g_settings->getBool("desynchronize_mapblock_texture_animation"))
+			//	tile_info.anim.frame_offset =
+			//			100000 * (2.0 + noise3d(
+			//			blockpos.X, blockpos.Y,
+			//			blockpos.Z, 0));
+
+			for (auto &frame : *layer->frames)
+				tile_info.anim.frames.push_back(frame.texture);
+		}
+
+		layer->atlas_tile_info_index = m_tiles_infos.size();
+		m_tiles_infos.push_back(tile_info);
+		unique_layers.push_back(layer);
+
+		atlas_size += (tile_info.width * tile_info.height);
+	}
+
+	u32 atlas_side = std::min(getClosestPowerOfTwo(std::sqrt((f32)atlas_size)), 16384u);
+
+	packTextures(atlas_side);
+
+	// Extending the atlas width twice for allocating the additional texture space for cracks
+	m_atlas_texture = m_driver->addTexture(core::dimension2du(atlas_side*2, atlas_side), "DiffuseAtlas", video::ECF_A32B32G32R32F);
+
+	for (auto &info : m_tiles_infos)
+		if (info.tex)
+			m_atlas_texture->drawToSubImage(info.x, info.y, info.width, info.height, info.tex);
+
+	if (m_driver->getTextureCreationFlag(video::ETCF_CREATE_MIP_MAPS)) {
+		// Each tile in both sides of the atlas must have not less 2 pixels in a mip
+		u32 min_pixels_count = m_tiles_infos.size()*4;
+		m_atlas_texture->regenerateMipMapLevels(0, 0, min_pixels_count);
+	}
+}
+
+u32 TextureAtlas::getClosestPowerOfTwo(f32 num)
+{
+	return std::pow(2u, (u32)std::ceil(std::log2(num)));
+}
+
+bool TextureAtlas::canFit(const TileInfo &area, const TileInfo &tex)
+{
+	return (tex.width <= area.width && tex.height <= area.height);
+}
+
+void TextureAtlas::packTextures(int side)
```

Comment:
Could add a note here on what texture packing algorithm is used.

Reply:
Added the comment in the last commit.